src/__tests__/ceramic_integration.test.ts

```diff
@@ -0,0 +1,330 @@
+  const addressB = (await b.id()).addresses[0];
+  await a.swarm.connect(addressB);
+}
+
+async function makeCeramicCore(ipfs: IpfsApi, anchorServiceUrl: string | null, ethereumRpcUrl: string | null): Promise<Ceramic> {
+  const tmpFolder = await tmp.dir({ unsafeCleanup: true })
+  const gateway = !anchorServiceUrl
+  const ceramic = await Ceramic.create(ipfs, {networkName: 'local', pubsubTopic: TOPIC, stateStoreDirectory: tmpFolder.path, anchorServiceUrl, gateway, ethereumRpcUrl})
+  ceramic.did = makeDID()
+  await ceramic.did.authenticate()
+  return ceramic
+}
+
+function makeDID(): DID {
+  const seed = random.randomString(32);
+  const digest = sha256.hash(uint8arrays.fromString(seed));
+  const provider = new Ed25519Provider(digest);
+  const resolver = KeyDidResolver.getResolver();
+  return new DID({ provider, resolver });
+}
+
+async function makeGanache(startTime: Date, port: number): Promise<Ganache.Server> {
+  const ganacheServer = Ganache.server({
+    gasLimit: 7000000,
+    time: startTime,
+    mnemonic: 'move sense much taxi wave hurry recall stairs thank brother nut woman',
+    default_balance_ether: 100,
+    debug: true,
+    blockTime: 2,
+    network_id: 1337,
+    networkId: 1337,
+  });
+
+  await ganacheServer.listen(port);
+  return ganacheServer
+}
+
+interface MinimalCASConfig {
+  ipfsPort: number,
+  ceramicPort: number,
+  ganachePort: number,
+  mode: string,
+  port: number,
+}
+
+async function makeCAS(dbConnection: Connection, minConfig: MinimalCASConfig): Promise<CeramicAnchorApp> {
+  const configCopy = cloneDeep(config)
+  configCopy.mode = minConfig.mode
+  configCopy.port = minConfig.port
+  configCopy.anchorControllerEnabled = true
+  configCopy.ipfsConfig.url = 'http://localhost:' + minConfig.ipfsPort
+  configCopy.ipfsConfig.pubsubTopic = TOPIC
+  configCopy.ceramic.apiUrl = 'http://localhost:' + minConfig.ceramicPort
+  configCopy.blockchain.connectors.ethereum.network = "ganache"
+  configCopy.blockchain.connectors.ethereum.rpc.port = minConfig.ganachePort + ''
+  const childContainer = container.createChildContainer()
+  return new CeramicAnchorApp(childContainer, configCopy, dbConnection)
+}
+
+export async function anchorUpdate(stream: Stream, anchorService: CeramicAnchorApp): Promise<void> {
+  // The anchor request is not guaranteed to already have been sent to the CAS when the create/update
+  // promise resolves, so we wait a bit to give the ceramic node time to actually send the request
+  // before triggering the anchor.
+  // TODO: Remove this once Ceramic won't return from a request that makes an anchor without having
+  // already made the anchor request against the CAS.
+  await Utils.delay(500)
+
+  const tillAnchored = stream
+    .pipe(
+      filter((state) => [AnchorStatus.ANCHORED, AnchorStatus.FAILED].includes(state.anchorStatus)),
+      take(1),
+    )
+    .toPromise();
+  await anchorService.anchor();
+  await tillAnchored;
```

Comment:
jw is there a reason `tillAnchored` should be constructed before the call to `anchorService.anchor` is made?

Reply:
yes, have to have the listener registered before the anchor or else the anchor might complete before you start listening and then you wouldn't get any updates

Reply:
Oh wait, actually this isn't quite true anymore. Now that streams get fixed to the version they are loaded at, then even if the anchor had completed before you subscribed to changed you'd still get notified.  I can switch things around now that that's the case so the control flow is simpler